### PR TITLE
Fix for #3621 freeze in gtextfield_mouse

### LINKED ISSUE #3621: Freeze! Execute script dialog, paste, try to double click to select line, freeze in gtextfield_mouse
Arch Linux, https://github.com/fontforge/fontforge/commit/26a0b528801d2e1281a009b8a36338124eb6f7c1

Reproduction steps:
1. File &rarr; Execute Script&hellip;
1. Write multiple lines, for example `test\ntest\ntest` (Where `\n` is a new line.)
2. Paste something long that goes over one line between the first and second line. For example, try pasting `test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test/test`.
3. As in #3620, keep clicking on the new line you pasted. Click click click, and drag a bit while clicking sometimes.
4. You'll get a freeze!

On killing the program, you'll get this:
```
Program received signal SIGINT, Interrupt.
0x0000155555073754 in gtextfield_mouse (g=0x555555dc9530, event=0x7fffffffae30)
    at gtextfield.c:1761
1761                for ( j=0; gt->lines[i+1]!=-1 && gt->sel_base>=gt->lines[i+1]; ++j );
(gdb) bt
#0  0x0000155555073754 in gtextfield_mouse
    (g=0x555555dc9530, event=0x7fffffffae30) at gtextfield.c:1761
#1  0x00001555550130ab in _GWidget_Container_eh
    (gw=0x555555e07ae0, event=0x7fffffffae30) at gcontainer.c:302
#2  0x0000155555014f1e in _GWidget_TopLevel_eh
    (gw=0x555555e07ae0, event=0x7fffffffae30) at gcontainer.c:745
#3  0x000015555508fcd2 in _GGDKDraw_CallEHChecked
    (gw=0x555555e07ae0, event=0x7fffffffae30, eh=
    0x155555014590 <_GWidget_TopLevel_eh>) at ggdkdraw.c:336
#4  0x00001555550924f0 in _GGDKDraw_DispatchEvent
    (event=0x555555646e60, data=0x555555662320) at ggdkdraw.c:1211
#5  0x000015555384ac86 in  () at /usr/lib/libgdk-3.so.0
#6  0x000015555387cab5 in  () at /usr/lib/libgdk-3.so.0
#7  0x0000155552a3f7bf in g_main_context_dispatch ()
    at /usr/lib/libglib-2.0.so.0
#8  0x0000155552a41739 in  () at /usr/lib/libglib-2.0.so.0
#9  0x0000155552a4177e in g_main_context_iteration ()
    at /usr/lib/libglib-2.0.so.0
#10 0x0000155555095096 in GGDKDrawProcessOneEvent (gdisp=0x555555662320)
    at ggdkdraw.c:2181
#11 0x00001555550183a3 in GDrawProcessOneEvent (gdisp=0x555555662320)
    at gdraw.c:767
#12 0x000015555536118a in ScriptDlg (fv=0x5555558b8db0, cv=0x0)
    at scriptingdlg.c:360
#13 0x000015555528526d in FVMenuExecute
    (gw=0x555555d432b0, UNUSED_mi=0x555555d463e0, UNUSED_e=0x7fffffffbdb0)
    at fontview.c:1227
#14 0x000015555503ae4b in gmenu_mouse (m=0x555555d808e0, event=0x7fffffffbdb0) at gmenu.c:943
#15 0x000015555503c912 in gmenu_eh (w=0x555555ebfc10, ge=0x7fffffffbdb0) at gmenu.c:1384
#16 0x000015555508fcd2 in _GGDKDraw_CallEHChecked (gw=0x555555ebfc10, event=0x7fffffffbdb0, eh=0x15555503c839 <gmenu_eh>) at ggdkdraw.c:336
#17 0x00001555550924f0 in _GGDKDraw_DispatchEvent (event=0x555555e370a0, data=0x555555662320) at ggdkdraw.c:1211
#18 0x000015555384ac86 in  () at /usr/lib/libgdk-3.so.0
#19 0x000015555387cab5 in  () at /usr/lib/libgdk-3.so.0
#20 0x0000155552a3f7bf in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#21 0x0000155552a41739 in  () at /usr/lib/libglib-2.0.so.0
#22 0x0000155552a4177e in g_main_context_iteration () at /usr/lib/libglib-2.0.so.0
#23 0x0000155555095112 in GGDKDrawEventLoop (gdisp=0x555555662320) at ggdkdraw.c:2192
#24 0x0000155555018486 in GDrawEventLoop (gdisp=0x555555662320) at gdraw.c:787
#25 0x00001555553a17bc in fontforge_main (argc=2, argv=0x7fffffffe1d8) at startui.c:1432
#26 0x0000555555555159 in main (argc=2, argv=0x7fffffffe1d8) at main.c:39
```

### PR DESCRIPTION
**Pull request by skef**
I don't know what the intention of the old line-selection code was. This replacement selects the semantic rather than the visible line, which matches most other implementations I'm familiar with.

Closes #3621 

I also removed one poor initialization of `sel_end`, but it shouldn't be related to #3620. 

If a newline char value can be the latter half of a wide char then this could have undesired results, but it shouldn't fail -- it might just select less than a semantic line if the wide char in question falls at the end of a visible line. 

### Types of changes
What types of changes does your code introduce? Check all the boxes that apply:
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist
Go over all the following points and check all the boxes that apply. 
If you're unsure about any of these, don't hesitate to ask. We're here to help! Various areas of the codebase have been worked on by different people in recent years, so if you are unfamiliar with the general area you're working in, please feel free to chat with people who have experience in that area. See the list [here](https://github.com/fontforge/fontforge/blob/master/CONTRIBUTING.md#people-to-ask).
- [x] My code follows the code style of this project found [here](https://github.com/fontforge/fontforge/blob/master/CONTRIBUTING.md#coding-style).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/fontforge/fontforge/blob/master/CONTRIBUTING.md) guidelines.

### CODE REVIEW

**Comment by skef** [comment]
Now my wrist hurts ...

File: gdraw/gtextfield.c [hunk 1]
```diff
@@ -1754,14 +1755,15 @@ return( true );
...
 	} else if ( gt->linesel ) {
-	    int j, e;
-	    gt->sel_start = gt->lines[i]; gt->sel_end = gt->lines[i+1];
-	    if ( gt->sel_end==-1 ) gt->sel_end = u_strlen(gt->text);
-	    for ( j=0; gt->lines[i+1]!=-1 && gt->sel_base>=gt->lines[i+1]; ++j );
-	    if ( gt->sel_start<gt->lines[i] ) gt->sel_start = gt->lines[i];
-	    e = gt->lines[j+1]==-1 ? u_strlen(gt->text): gt->lines[j+1];
-	    if ( e>gt->sel_end ) gt->sel_end = e;
+	    int j;
+	    for ( j=i; j>0 && gt->text[gt->lines[j]-1] != '\n'; --j )
```
Comment (jtanx) [inline gdraw/gtextfield.c]: I'm having a bit of trouble reproducing the original issue; anyway, would you mind giving a brief description of how this is hit/what this does?

**Comment by ctrlcctrlv** [comment]
Thanks for this @skef. @jtanx I'll make a video for you, I'm turning on my PC in a few minutes

**Comment by ctrlcctrlv** [comment]
Also @skef looking back you were right to reject my fix to this...because I just did the minimum so it would stop crashing, I didn't evaluate the code as you have done so expertly :)

File: gdraw/gtextfield.c (hunk 1 above)
Comment (skef) [reply gdraw/gtextfield.c]: I don't know exactly what the original code was supposed to do. Note that the old loop with `j` doesn't modify or test `j`. But conceptually what the code is supposed to do is clear: `linesel` is set to true on a triple-click and is supposed to select the line on which the click occurred. The intent of this new code is to select a "semantic" line -- a region of text between newlines -- rather than a visual line created by wrapping. There's no real reason to select the latter.

Comment (skef) [reply gdraw/gtextfield.c]: The issue, based on the symptoms, was that the guard on the loop was usually false, but when it was true you'd get a freeze for the self-evident reason.

**Comment by ctrlcctrlv** [comment]
Video of the freeze&mdash;notice how at the end I can no longer resize it. Same crash as in #3621 

[freeze.zip](https://github.com/fontforge/fontforge/files/3075595/freeze.zip)

**Comment by skef** [comment]
@ctrlcctrlv I haven't really done a clicking-all-over-the-place test of this new code. Only a moderate amount of clicking, with different lengths of lines and on the edge cases (first and last). I'm 75% confident that `gt->sel_base` is supposed to be the origin used for a shift-click, to select a region starting with the previous click. My guess is that the clicking occasionally results in a corrupt or just "weird" `gt->sel_base` value, and since that was in the guard of the loop that's what triggers the freeze. Anyway, the new code doesn't use it, or any of the initial `sel` values, so triple-clicking (at least) shouldn't cause that problem.

File: gdraw/gtextfield.c (hunk 1 above)
Comment (skef) [reply gdraw/gtextfield.c]: I just realized I could be a bit more helpful on this. 

`gt->lines[]` is an array of offsets into `gt->text` marking the start of a new *visual* line. It's a datastructure maintained in other code as contents are added and the window dimension changed.

**Comment by skef** [comment]
Thanks -- I should have checked that boundary condition.

**Comment by skef** [comment]
By the way -- the weird highlighting on a triple click is due to the use of one algorithm on mouse down and another on mouse up. That's an already-existing aspect of the code, and may be there for performance reasons; I'm not sure. Anyway, other than being slightly confusing it seems to be harmless.
